homeworks/solutions/torantie/2-ddg/src/halfedge.js

```diff
@@ -0,0 +1,387 @@
+            const firstHalfEdge = this.halfedge
+            const secondHalfEdge = firstHalfEdge.next
+            let i = 0
+
+            fn(firstHalfEdge.vertex, i)
+            i++;
+
+            for (let currentHalfEdge = secondHalfEdge; firstHalfEdge.vertex.idx !== currentHalfEdge.vertex.idx; currentHalfEdge = currentHalfEdge.next) {
+                fn(currentHalfEdge.vertex, i)
+                i++
+            }
+        } catch (e) {
+            console.error(e)
+        }
+    }
+
+    getNormalTriangle() {
+        let x = this.halfedge.getVector()
+        let y = this.halfedge.prev.twin.getVector()
+
+        return x.cross(y)
+    }
+
+}
+
+class Vertex {
+    constructor() {
+        this.position = null // Vector
+        this.halfedge = null // Halfedge
+        this.idx = -1   // Number
+    }
+
+    normal(method = 'equal-weighted') {
+        let sum = new Vector()
+
+        switch (method) {
+            case 'equal-weighted':
+                // TODO: compute euqally weighted normal of this vertex
+                this.forEachHalfEdge((currentHalfEdge) => {
+                    sum = sum.add(currentHalfEdge.face.getNormalTriangle());
+                })
+                break;
+            case 'area-weighted':
+                // TODO: compute area weighted normal of this vertex
+                this.forEachHalfEdge((currentHalfEdge) => {
+                    sum = sum.add(currentHalfEdge.face.getNormalTriangle().scale(currentHalfEdge.face.getAreaTriangle()));
+                })
+
+                break;
+            case 'angle-weighted':
+                // TODO: compute angle weighted normal of this vertex
+                this.forEachHalfEdge((currentHalfEdge) => {
+                    sum = sum.add(currentHalfEdge.face.getNormalTriangle().scale(currentHalfEdge.getAngle()));
+                })
+                break;
+            default: // undefined
+                return new Vector()
+        }
+
+        return sum.scale(1 / sum.norm())
+    }
+
+    curvature(method = 'Mean') {
+        switch (method) {
+            case 'Mean':
+                // TODO: compute mean curvature
+                return this.calculateMeanCurvature()
+            case 'Gaussian':
+                // TODO: compute Guassian curvature
+                return this.calculateGaussianCurvature()
+            case 'Kmin':
+                // TODO: compute principal curvature and return Kmin
+                return this.calculateK(1)
+            case 'Kmax':
+                // TODO: compute principal curvature and return Kmax
+                return this.calculateK(2)
+            default: // undefined
+                return 0
+        }
+    }
+
+    // NOTE: you can add more methods if you need here
+
+    calculateK(num) {
+        let h = this.calculateMeanCurvature()
+        let k = this.calculateGaussianCurvature()
+
+        switch (num) {
+            case 1:
+                return h - Math.sqrt(Math.pow(h, 2) - k)
```

Comment:
Theoretically, h^2-k is always non-negative. But in practice, this can fall into negative values (because of bad triangulation/discretization).
Usually, we handle the case by falling back to mean curvature, that is `h^2-k < 0 ? 0 : h^2-k`